drivers: dp: add STM32 support

fabiobaltieri opened the pull request:
Add support for direct control of STM32 gpios to the DP driver.

---

Tested on an  F3 and on an H7, should work on all of them.

fabiobaltieri committed:
drivers: dp: add STM32 support

Add support for direct control of STM32 gpios to the DP driver.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented on drivers/dp/swdp_ll_pin.h:
```diff
@@ -7,14 +7,18 @@
 #define CPU_CLOCK		64000000U
 #else
 #define CPU_CLOCK		CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
 #endif
 
-#if defined(CONFIG_SOC_SERIES_NRF52X) || defined(CONFIG_SOC_SERIES_NRF53X)
+#if defined(CONFIG_SOC_SERIES_NRF52X) || defined(CONFIG_SOC_SERIES_NRF53X) || \
+	defined(CONFIG_SOC_FAMILY_STM32)
```
@jfischer-no this strategy could get real messy real quick, want me to refactor this into multiple files? Like have an `swdp_ll_noop.h`, `swdp_ll_nrf.h`, `swdp_ll_stm32.h` and include them selectively from this fie.

keith-zephyr replied:
+1 creating architecture specific headers and then defining a macro/helper in each header to eliminate the #ifdefs in this file is much easier to read.

fabiobaltieri replied:
done, way cleaner now

fabiobaltieri committed:
drivers: dp: move the nrf code to its own file

Move the nrf specific functions to a separate file so that every soc has
a dedicated header.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
Okay did a bit of refactoring, platforms have specific headers now, only one guard in the main file.

etienne-lms commented on drivers/dp/swdp_ll_pin_stm32.h:
```diff
@@ -0,0 +1,51 @@
+{
+	GPIO_TypeDef *gpio = base;
+
+	LL_GPIO_SetPinMode(gpio, BIT(pin), LL_GPIO_MODE_INPUT);
```
I think these partial write accesses in GPIO regrsiters should be protected by a semaphore.
Could we use here the semaphore used in **gpio_stm32_.c**?
```c
	z_stm32_hsem_lock(CFG_HW_GPIO_SEMID, HSEM_LOCK_DEFAULT_RETRY);
	(...)
	z_stm32_hsem_unlock(CFG_HW_GPIO_SEMID);

fabiobaltieri commented:
Added hsem locking, for multi-core devices, same as `gpio_stm32.c`

jfischer-no commented on drivers/dp/swdp_ll_pin_stm32.h:
```diff
@@ -0,0 +1,68 @@
+	GPIO_TypeDef *gpio = base;
+	uint32_t val;
+
+	z_stm32_hsem_lock(CFG_HW_GPIO_SEMID, HSEM_LOCK_DEFAULT_RETRY);
+
+	val = LL_GPIO_ReadOutputPort(gpio);
+	val |= BIT(pin);
+	LL_GPIO_WriteOutputPort(gpio, val);
+
+	z_stm32_hsem_unlock(CFG_HW_GPIO_SEMID);
```
Is it significantly faster than just GPIO API? I guess z_stm32_hsem_lock/unlock will be optimized away on single core?

fabiobaltieri replied:
Yes and yes, I think about 250khz vs 550khz on the C0, something like 800 on the H7 can't remember right now but yeah big difference. Lock and lock are indeed no-ops on single cores.

jfischer-no reviewed (approved):
Thanks

mathieuchopstm commented on drivers/dp/swdp_ll_pin_stm32.h:
```diff
@@ -0,0 +1,68 @@
+
+	z_stm32_hsem_lock(CFG_HW_GPIO_SEMID, HSEM_LOCK_DEFAULT_RETRY);
+
+	val = LL_GPIO_ReadOutputPort(gpio);
+	val |= BIT(pin);
+	LL_GPIO_WriteOutputPort(gpio, val);
```
`LL_GPIO_SetOutputPin` is a single write and performs RMW atomically in hardware. Am I missing some context, or couldn't it be used here?

fabiobaltieri replied:
Oh, brilliant, so no lock needed. Thanks for pointing that out I'll send a followup to use that instead and drop the locks.

fabiobaltieri replied:
Nothing is ever easy https://github.com/zephyrproject-rtos/zephyr/pull/90211#issuecomment-2893795131

mathieuchopstm commented on drivers/dp/swdp_ll_pin_stm32.h:
```diff
@@ -0,0 +1,68 @@
+
+	z_stm32_hsem_lock(CFG_HW_GPIO_SEMID, HSEM_LOCK_DEFAULT_RETRY);
+
+	val = LL_GPIO_ReadOutputPort(gpio);
+	val &= ~BIT(pin);
+	LL_GPIO_WriteOutputPort(gpio, val);
```
Ditto but with `LL_GPIO_ResetOutputPin` instead.

mathieuchopstm reviewed (commented):
FYI @fabiobaltieri
